src/compile/data/data.ts

```diff
@@ -149,46 +270,100 @@ export function parseLayerData(model: LayerModel): DataComponent {
 export function assembleData(model: Model, data: VgData[]) {
   const component = model.component.data;
 
-  const sourceData = source.assemble(model, component);
-  if (sourceData) {
-    data.push(sourceData);
+  if (!component || !component.source) {
+    return data;
   }
 
-  summary.assemble(component, model).forEach(function(summaryData) {
-    data.push(summaryData);
-  });
+  let dataSource = source.assemble(model, component);
+  data.push(dataSource);
+
+  dataSource.transform = [].concat(
+    calculate.assemble(component),
+    filter.assemble(component),
+    nullFilter.assemble(component),
+    bin.assemble(component),
+    timeUnit.assemble(component)
+  );
+
+  let rawSource;
+  const assembleRaw = component.aggregate && component.aggregate.assembleRaw;
+  if (assembleRaw) {
+    // create an intermediate data source for raw data
+    const sourceName = dataSource.name;
+    const rawName = model.dataName(RAW);
+    rawSource = dataSource;
+    dataSource.name = rawName;
+
+    dataSource = {
+      source: rawName,
+      name: sourceName
+    };
+    data.push(dataSource);
+  }
+
+  const aggregates = aggregate.assemble(component, model);
+  if (aggregates.length > 0) {
+    dataSource.transform = (dataSource.transform || []).concat(aggregates);
+  }
 
-  if (data.length > 0) {
-    const dataTable = data[data.length - 1];
+  // add rank transform
+  const rank = colorRank.assemble(component);
+  if (rank.length > 0) {
+    dataSource.transform = (dataSource.transform || []).concat(rank);
+  }
+
+  if (empty(dataSource.transform)) {
+    delete dataSource.transform;
+  }
+
+  // stack
+  const stackTransform = stackScale.assemble(component);
+  if (stackTransform) {
+    data.push({
+      name: model.dataName(STACKED_SCALE),
+      source: dataSource.name,
+      transform: [stackTransform]
+    });
+  }
 
-    // color rank
-    const colorRankTransform = colorRank.assemble(component);
-    if (colorRankTransform.length > 0) {
-      dataTable.transform = (dataTable.transform || []).concat(colorRankTransform);
+  // scale
+  const posFilter = nonPositiveFilter.assemble(component);
+  if (posFilter) {
+    if (assembleRaw) {
+      rawSource.transform = (rawSource.transform || []).concat(posFilter);
+    } else {
+      dataSource.transform = (dataSource.transform || []).concat(posFilter);
     }
+  }
+
+  // time unit domain
+  const tuDomain = timeUnitDomain.assemble(component);
+  if (tuDomain) {
+    data.push.apply(data, tuDomain);
+  }
+
+  // remove data sources that are not needed
```

Comment:
Is it possible to just not add them at all up front?  removing later is generally a bad pattern that we should avoid unless we really need to.   Plus, you are iterating over the entire `data` array that is an input parameter, which means you might remove data source that's not even generated by this `assembleData`.  This seems like a very risky pattern.  

I guess you're only adding this to remove raw?   I think even if raw doesn't have its own transform.  
There is a good reason that we might need it separately.  
For example, if source is aggregated by `O1` and `O2`.  To sort ordinal scale of `O1`, you still need the raw source even if you don't do any transform in raw. 

Even if I'm wrong about we shouldn't remove, it seems like if we move the posFilter up to before adding the raw one, we can check if data source already contains raw.